wgengine/tstun/tun.go

```diff
@@ -172,19 +205,33 @@ func (t *TUN) poll() {
 }
 
 func (t *TUN) filterOut(buf []byte) filter.Response {
+	p := t.parsedPacketPool.Get().(*packet.ParsedPacket)
+	p.Decode(buf)
+
+	if t.PreFilterOut != nil {
+		if t.PreFilterOut(p, t) == filter.Drop {
+			return filter.Drop
+		}
+	}
+
 	filt, _ := t.filter.Load().(*filter.Filter)
 
 	if filt == nil {
-		t.logf("Warning: you forgot to use SetFilter()! Packet dropped.")
+		t.logf("tstun: warning: you forgot to use SetFilter()! Packet dropped.")
+		return filter.Drop
+	}
+
+	if filt.RunOut(p, t.filterFlags) != filter.Accept {
 		return filter.Drop
 	}
 
-	var p packet.ParsedPacket
-	if filt.RunOut(buf, &p, t.filterFlags) == filter.Accept {
-		return filter.Accept
+	if t.PostFilterOut != nil {
+		if t.PostFilterOut(p, t) == filter.Drop {
+			return filter.Drop
+		}
 	}
 
-	return filter.Drop
+	return filter.Accept
```

Comment:
Changing the default to accept worries me, it makes it easier for a future change to accidentally fail open instead of closed.

Given the code structure, I'm not sure how to change it so that we can default to Drop, so I guess this comment is just me being worried, nothing actionable :/

Reply:
The tstun tests should catch at least obvious mistakes (such as bypassing the main filter entirely).